src/main/java/com/faforever/client/login/OAuthValuesReceiver.java

```diff
@@ -0,0 +1,134 @@
+  private static final Pattern STATE_PATTERN = Pattern.compile("state=([^ &]+)");
+  private static final List<String> ALLOWED_HOSTS = List.of("localhost");
+
+  private final CompletableFuture<URI> addressFuture = new CompletableFuture<>();
+
+  private final ClientProperties clientProperties;
+  private final PreferencesService preferencesService;
+
+  public CompletionStage<Values> receiveValues() {
+    return CompletableFuture.supplyAsync(() -> {
+      Collection<URI> redirectUris = retrieveRedirectUris();
+
+      for (URI uri : redirectUris) {
+        try {
+          return readWithUri(uri);
+        } catch (SocketException e) {
+          log.warn("Port " + uri.getPort() + " is probably already in use", e);
+        }
+      }
+      throw new IllegalStateException("Could not read from any redirect URI: " + redirectUris);
+    });
+  }
+
+  @SneakyThrows
+  private List<URI> retrieveRedirectUris() {
+    JavaFxUtil.assertBackgroundThread();
+    if (clientProperties.isUseRemotePreferences()) {
+      return preferencesService.getRemotePreferences().getOAuth().getRedirectUris().stream()
+          .filter(uri -> uri.getPort() > 0)
+          .filter(uri -> ALLOWED_HOSTS.contains(uri.getHost()))
+          .collect(Collectors.toList());
+    }
+    return IntStream.range(60000, 65535)
+        .mapToObj(port -> URI.create("http://localhost:%d/".formatted(port)))
+        .collect(Collectors.toList());
+  }
+
+  @SneakyThrows
+  private Values readWithUri(URI uri) throws SocketException {
+    // A random port can't be used since the redirect URI, including port, must be registered on the server
+    try (ServerSocket serverSocket = new ServerSocket(uri.getPort(), 1, InetAddress.getLoopbackAddress())) {
+      addressFuture.complete(uri);
+
+      Socket socket = serverSocket.accept();
+      Values values = readValues(socket);
+      writeResponse(socket);
+      return values;
+    }
+  }
+
+  private static void writeResponse(Socket socket) throws IOException {
+    try (Writer writer = new BufferedWriter(new OutputStreamWriter(socket.getOutputStream()))) {
+      @SuppressWarnings("ConstantConditions")
+      String html = new String(OAuthValuesReceiver.class.getResourceAsStream("/login_success.html").readAllBytes());
+
+      writer
+          .append("HTTP/1.1 200 OK\r\n")
+          .append("Content-Length ")
+          .append(String.valueOf(html.length()))
+          .append("\r\n")
+          .append("Content-Type: text/html\r\n")
+          .append("Connection: Closed\r\n")
+          .append("\r\n")
+          .append(html);
+    }
+  }
+
+  private Values readValues(Socket socket) throws IOException {
+    // Don't use try-with-resources since socket must not be closed yet
+    BufferedReader reader = new BufferedReader(new InputStreamReader(socket.getInputStream()));
+    String query = reader.readLine();
+    String code = extractValue(query, CODE_PATTERN);
```

Comment:
If this is just parsing the URL wouldn't it be simpler to just use the URLEncodeUtils to parse the url and query parameters rather than have the custom regex

Reply:
Agree, however, `URLEncodeUtils` doesn't fit since it's a client-side library to "encode" URLs whereas this is server-side parsing of an HTTP request: `GET /?code=1234&state=abcd HTTP/1.1`. I'm not aware of any library that can easily parse this. I could use Spring classes to start an embedded Tomcat or so but that seems to be total overkill.